### PR TITLE
extmod/modwebsocket: Add client support.

### PR DESCRIPTION
**Pull request by ccooper21**
This change adds websocket client support as per RFC6455.  Section 5.1 states
that a websocket client must mask all frames that it sends to the server.
Further, section 5.3 states the masking key is a 32-bit value chosen at random
by the client unique to each frame.  These requirements are now implemented.
A non-compliant debug mask is also supported, allowing control of the masking
key.

### CODE REVIEW

**Comment by pfalcon** [comment]
Thanks for this contribution! What's interesting and that I lately also was thinking how to "finish" this module and in particular add client support. But I thought also about other things, and it makes it not so easy... Let me however start with pointing few immediate problems with this patch as it is.

File: extmod/modwebsocket.c [hunk 1]
```diff
@@ -58,16 +63,39 @@ typedef struct _mp_obj_websocket_t {
 STATIC mp_uint_t websocket_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode);
 
 STATIC mp_obj_t websocket_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
-    mp_arg_check_num(n_args, n_kw, 1, 2, false);
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_sock, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: This should be empty qstr.

File: extmod/modwebsocket.c [hunk 2]
```diff
@@ -58,16 +63,39 @@ typedef struct _mp_obj_websocket_t {
...
+        { MP_QSTR_use_blocking_writes, MP_ARG_BOOL, {.u_bool = false} },
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: We always save for the space, so using such long arg names is no-no.

File: extmod/modwebsocket.c [hunk 3]
```diff
@@ -58,16 +63,39 @@ typedef struct _mp_obj_websocket_t {
...
+        { MP_QSTR_is_client, MP_ARG_BOOL, {.u_bool = false} },
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: This should use the same terminology as "ssl" module.

File: extmod/modwebsocket.c [hunk 4]
```diff
@@ -58,16 +63,39 @@ typedef struct _mp_obj_websocket_t {
...
+        { MP_QSTR_debug_mask, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: I'm always for having debugging/diagnostic facilities in the code, but they should be properly #ifdef'ed and off by default.

File: extmod/modwebsocket.c [hunk 5]
```diff
@@ -58,16 +63,39 @@ typedef struct _mp_obj_websocket_t {
...
+        mp_arg_val_t sock, use_blocking_writes, is_client, debug_mask;
+    } arg_vals;
+    mp_arg_parse_all_kw_array(n_args, n_kw, args,
+        MP_ARRAY_SIZE(allowed_args), allowed_args, (mp_arg_val_t*)&arg_vals);
+
     mp_obj_websocket_t *o = m_new_obj(mp_obj_websocket_t);
     o->base.type = type;
-    o->sock = args[0];
+    o->sock = arg_vals.sock.u_obj;
+    o->do_write_masking = !arg_vals.is_client.u_bool ? NO_WRITE_MASKING : NORMAL_WRITE_MASKING;
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: At least remove "!" and swap args. I'd also use normal "if" here.

File: extmod/modwebsocket.c [hunk 6]
```diff
@@ -229,6 +257,34 @@ STATIC mp_uint_t websocket_write(mp_obj_t self_in, const void *buf, mp_uint_t si
...
+            // RFC6455 Section 5.3 states that the masking key must be derived
+            // from a strong source of entropy.  The "urandom" module doesn't
+            // qualify in this regard, but there isn't any cross-platform
+            // alternative.  Fortunately, the purpose of masking is not
+            // cryptographically motivated.  The "urandom" module should be
+            // seeded though, otherwise upon restart, the same sequence of
+            // masks will always be used.  A seed could be derived from a
+            // network resource, a network interface's characteristics or
+            // statistics, or a platform specific resource.  Examples of using
+            // a platform specific resource include reading an ESP8266's
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: I'm all for detailed comments, but specifics of a particular chip in a generic module are superfluous.

File: extmod/modwebsocket.c [hunk 7]
```diff
@@ -229,6 +257,34 @@ STATIC mp_uint_t websocket_write(mp_obj_t self_in, const void *buf, mp_uint_t si
...
+    if (self->do_write_masking != NO_WRITE_MASKING) {
+        hdr_sz += 4;
+        header[1] |= 0x80;
+        if (self->do_write_masking == NORMAL_WRITE_MASKING) {
+
+            // RFC6455 Section 5.3 states that the masking key must be derived
+            // from a strong source of entropy.  The "urandom" module doesn't
+            // qualify in this regard, but there isn't any cross-platform
+            // alternative.  Fortunately, the purpose of masking is not
+            // cryptographically motivated.  The "urandom" module should be
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: I'd remove such a statement, it's subjective.

File: extmod/modwebsocket.c [hunk 8]
```diff
@@ -229,6 +257,34 @@ STATIC mp_uint_t websocket_write(mp_obj_t self_in, const void *buf, mp_uint_t si
...
+            // seeded though, otherwise upon restart, the same sequence of
+            // masks will always be used.  A seed could be derived from a
+            // network resource, a network interface's characteristics or
+            // statistics, or a platform specific resource.  Examples of using
+            // a platform specific resource include reading an ESP8266's
+            // 32-bit Random Number Generator register, or reading consecutive
+            // values from a floating analog pin.
+            mp_obj_t dest[3];
+            mp_load_method(mp_module_get(MP_QSTR_urandom), MP_QSTR_getrandbits, dest);
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: The code to get a random value should be factored out to a function.

File: extmod/modwebsocket.c [hunk 9]
```diff
@@ -239,7 +295,15 @@ STATIC mp_uint_t websocket_write(mp_obj_t self_in, const void *buf, mp_uint_t si
 
     mp_uint_t out_sz = mp_stream_write_exactly(self->sock, header, hdr_sz, errcode);
     if (*errcode == 0) {
-        out_sz = mp_stream_write_exactly(self->sock, buf, size, errcode);
+        if (self->do_write_masking == NO_WRITE_MASKING) {
+            out_sz = mp_stream_write_exactly(self->sock, buf, size, errcode);
+        } else {
+            byte masked_buf[size];
```
Comment (pfalcon) [inline extmod/modwebsocket.c]: Really? But size can be 1GB.

**Review by pfalcon — CHANGES_REQUESTED** [review]

**Comment by pfalcon** [comment]
Besides the above the bigger conceptual problem is how to support non-blocking write mode for such stream wrappers: https://github.com/micropython/micropython/issues/3396 . I invite to think how the code you add can be made useful for async non-blocking writers. (If you don't have ideas/not interested, I can share some of my ideas.)

**Comment by pfalcon** [comment]
Async implementation of websockets which needs client support: https://github.com/micropython/micropython-lib/pull/228

**Comment by ccooper21** [comment]
@pfalcon Thanks for the quick review, feedback, and encouragement!  I hope to get this wrapped up in short order and move on to further tasks.  To provide context, I'm working on integrating with the AWS IoT MQTT-based service via websockets on an ESP8266.  I have this working on the UNIX port, but need to do further work for the ESP8266 port due to the MCU's constraints.  Ultimately this will require submitting more pull requests, and likely revisiting some stagnant pull requests from others.  

I think I am aware of the considerations related to writing code for an embedded environment, but I am finding translating this into specific decisions to be more challenging.  Most of my experience is related to enterprise software, which has a completely different set of considerations.  I will likely have some questions regarding your feedback shortly.  Thanks for your guidance as I proceed!

**Comment by dpgeorge** [comment]
Closing due to inactivity (specifically, the feedback has not been addressed).

This would be a great feature, so feel free to reopen against latest master.  Things to consider would be how to make it work with uasyncio (ie non-blocking).

**Pull request closed without merging**